src/main/java/org/pkcs11/jacknji11/CryptokiE.java

```diff
@@ -1873,4 +1857,45 @@ private byte[] resize(byte[] buf, int newSize) {
         System.arraycopy(buf, 0, result, 0, result.length);
         return result;
     }
+
+    public Metrics getMetrics() {
+        return c.getMetrics();
+    }
+
+    /**
+     * Set the strategy to use for getting the length of an attribute.
+     *
+     * @param getAttributeLengthStrategy the strategy to use for getting the length of an attribute
+     */
+    public void setQueryLengthStrategy(GetAttributeLengthStrategy getAttributeLengthStrategy) {
```

Comment:
I would not change it per call. In case of multi-threaded apps it would be rather difficult to manage. I think the objective is met (reducing number of fetches) by default implementation pretty well. I'm just following advice to make things configurable and using " .... different settings per instance ...".

Reply:
I think two implementations are currently possible:

1. IndefiniteSize (use length=0 for every attr) - this was previous behavior of the library.
2. MaxSize (use some max length depending on attr) which I implemented in a "Default" impl

It is probably up to you to tell me which one should be the actual default. 
And I would rename the "Default" to "MaxSize" since it conveys what it actually does.

Reply:
Today, previous to your PR, IndefiniteSize would be the "default" right? In that case it may be worth keeping that in order to avoid unexpected behavior change. But making it easy to select MaxSize.
Good idea to rename Default imho.

We would have to do extensive testing on MaxSize, on many different cryptographic modules, in order to make it the default safely without chance of regression. It's probably too much work to test out.

I understand the multi-threading issue. 

Just for the sake of discussion. Perhaps something for the future, actually being able to set size values if you know what size the attributes you're looking for should be? Typically I make calls reading say all attributes for an RSA key, and different ones for an EC key, or PQC keys, based on what KEY_TYPE they report. Typically reading one attribute first, and then the rest in a batch. (Trying to speculatively read all different key attributes gets us into CKR_ATTRIBUTE_TYPE_INVALID on some HSMs, and seems quite slow...)

Reply:
Thanks for the detailed discussion. I do like the idea. It's easier to gather the optimization complexity in the library than to re-implement it everywhere. Also the code is nice imho.
I can run basic tests on a range of different HSMs once the review is done. 

This thread can be resolved if @joelhockey agrees, you have the final say on inclusion.